src/python/pants/engine/goal.py

```diff
@@ -13,6 +13,10 @@
 from pants.util.meta import classproperty
 
 
+if TYPE_CHECKING:
+  from pants.engine.rules import UnionMembership
```

Comment:
Circular dependency :/

Reply:
Would it be reasonable to move `UnionMembership` out of `rules.py`? That file is pretty large anyway.

Reply:
I think so - would this be `pants.engine.unions`?